packages/core/utils/src/dml/helpers/entity-builder/infer-primary-key-properties.ts

```diff
@@ -0,0 +1,74 @@
+import { DMLSchema } from "../../entity-builder"
+import { IdProperty } from "../../properties/id"
+
+/*
+  The id() property is an core opinionated property that will act as a primaryKey
+  by default and come with built-in logic when converted to a mikroorm entity. If no other
+  primaryKey() properties are found within the schema, we continue treating the id() property 
+  as a primaryKey. When other fields are set as explicit primaryKey fields, we convert the 
+  id() property to no longer be a primaryKey. 
+
+  Example:
+  Model 1:
+    id: model.id() -> primary key
+    code: model.text()
+
+  Model 2:
+    id: model.id()
+    code: model.text().primaryKey() -> primary key
+
+  Model 3:
+    id: model.id()
+    code: model.text().primaryKey() -> composite primary key
+    name: model.text().primaryKey() -> composite primary key
+*/
+export function inferPrimaryKeyProperties<TSchema extends DMLSchema>(
+  schema: TSchema
+) {
+  // If explicit primaryKey fields are not found, no inferrence is required. Return early.
+  if (!getExplicitPrimaryKeyFields(schema).length) {
+    return schema
+  }
+
+  // If explicit primaryKey fields are found, set any id() properties to no longer be
+  // set to primaryKey.
+  for (const [field, property] of Object.entries(schema)) {
+    const parsed = property.parse(field)
+    const isRelationshipType = "type" in parsed
+
+    if (isRelationshipType) {
+      continue
+    }
+
+    if (parsed.dataType.name === "id") {
```

Comment:
Nit: for consistency and make it easier can you use the same approach as the other ones and if there is no type you can add a symbol.for to which we can refer too without being shown in the auto completion and then we can use it every time we need without having to think about which value to compare to know that it is an id.

You can look at the dml, properties and relations they all provide a static method ☺️

Reply:
I gave this a shot, but doesn't look nice when i tried it. Lets pair on this in the next PR. 👍🏻